src/Microsoft.AspNet.Mvc.Core/Routing/AttributeRoutePrecedence.cs

```diff
@@ -32,13 +32,66 @@ public static decimal Compute(RouteTemplate template)
             return precedence;
         }
 
+        public static decimal ComputeGenerated(RouteTemplate template)
+        {
+            // Each precedence digit corresponds to one decimal place. For example, 3 segments with precedences 2, 1,
+            // and 4 results in a combined precedence of 2.14 (decimal).
+            var precedence = 0m;
+
+            for (var i = 0; i < template.Segments.Count; i++)
+            {
+                var segment = template.Segments[i];
+
+                var digit = ComputeGenerationDigit(segment);
+                Debug.Assert(digit >= 0 && digit < 10);
+
+                precedence += decimal.Divide(digit, (decimal)Math.Pow(10, i));
+            }
+
+            return precedence;
+        }
+
+        // Segments have the following order:
+        // 6 - Literal segments
+        // 5 - Multi-part segments
```

Comment:
I was lead by this comment: "// Multi-part segments should appear after literal segments but before parameter segments" which was already in the code to believe that we did want to treat multi-part segments and constrained parameters differently. If that's not the case I'll modify ComputeGenerationDigit and alter the existing comment so it doesn't contradict the code.